File: stations/hook-dispatcher/internal/handlers/pm_autopilot.go
package handlers

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/clients"
	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	entry := core.Entry{
		Handler:    pmAutopilotHandle,
		ModuleName: "pm_autopilot",
	}
	core.Register("SessionStart", entry)
	core.Register("PostToolUse", core.Entry{
		Matcher:    "Bash",
		Handler:    pmAutopilotHandle,
		ModuleName: "pm_autopilot",
	})
	core.Register("Stop", entry)
}

const (
	pmStateFile     = "/tmp/pm-autopilot-state.json"
	pmMaxSyncIssues = 3
)

var pmIssueRe = regexp.MustCompile(`(?:Closes|Fixes|Refs|Part of)\s+#(\d+)|(?:^|[^\w])#(\d+)\b`)
var pmBranchIssueRe = regexp.MustCompile(`#(\d+)`)

func pmRepo() string {
	// Prefer env var (easy override in tests)
	if r := os.Getenv("PM_AUTOPILOT_REPO"); r != "" {
		return r
	}
	// Read from config YAML directly — github.repo is a top-level key
	// that core.Config doesn't expose, so we read config files ourselves.
	root := os.Getenv("HOOK_OBSERVATORY_ROOT")
	if root == "" {
		home, _ := os.UserHomeDir()
		root = filepath.Join(home, "workshop", "stations", "hook-observatory")
	}
	for _, name := range []string{"config.yaml", "config.example.yaml"} {
		data, err := os.ReadFile(filepath.Join(root, name))
		if err != nil {
			continue
		}
		// Simple extraction: github:\n  repo: "owner/repo"
		lines := strings.Split(string(data), "\n")
		inGithub := false
		for _, line := range lines {
			trimmed := strings.TrimSpace(line)
			if trimmed == "github:" {
				inGithub = true
				continue
			}
			if inGithub {
				if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
					if strings.HasPrefix(trimmed, "repo:") {
						val := strings.TrimSpace(strings.TrimPrefix(trimmed, "repo:"))
						val = strings.Trim(val, `"'`)
						if val != "" {
							return val
						}
					}
				} else {
					inGithub = false
				}
			}
		}
	}
	return ""
}

func pmAutopilotHandle(eventType, toolName string, toolInput map[string]any, _ string) core.HookResult {
	repo := pmRepo()
	if repo == "" {
		return core.Allow()
	}
	switch eventType {
	case "SessionStart":
		return pmSessionStart(repo)
	case "PostToolUse":
		if toolName == "Bash" {
			return pmPostBash(toolInput, repo)
		}
	case "Stop":
		return pmOnStop()
	}
	return core.Allow()
}

// ---------------------------------------------------------------------------
// A. SessionStart
// ---------------------------------------------------------------------------

func pmSessionStart(repo string) core.HookResult {
	out, err := clients.RunGH([]string{
		"issue", "list", "--state", "open",
		"--json", "number,title,labels",
		"--limit", "20", "--repo", repo,
	})
	if err != nil {
		pmSaveState(map[string]any{"issues": []any{}, "in_progress": []any{}, "ready": []any{}})
		return core.Allow()
	}

	var issues []map[string]any
	if err := json.Unmarshal([]byte(out), &issues); err != nil || len(issues) == 0 {
		pmSaveState(map[string]any{"issues": []any{}, "in_progress": []any{}, "ready": []any{}})
		return core.Allow()
	}

	var inProgress, blocked, ready []map[string]any
	for _, issue := range issues {
		labels := pmLabelNames(issue)
		switch {
		case pmHasLabel(labels, "in-progress"):
			inProgress = append(inProgress, issue)
		case pmHasLabel(labels, "blocked"):
			blocked = append(blocked, issue)
		default:
			ready = append(ready, issue)
		}
	}

	// Save state
	inProgressNums := pmNumbers(inProgress)
	readyNums := pmNumbers(ready)
	blockedNums := pmNumbers(blocked)
	pmSaveState(map[string]any{
		"issues":      issues,
		"in_progress": inProgressNums,
		"ready":       readyNums,
		"blocked":     blockedNums,
	})

	// Build markdown
	var parts []string
	parts = append(parts, "## GitHub PM Status")

	if len(inProgress) > 0 {
		parts = append(parts, "### In Progress")
		for _, i := range inProgress {
			num, _ := i["number"].(float64)
			title, _ := i["title"].(string)
			labels := pmLabelNames(i)
			var extra []string
			for _, l := range labels {
				if l != "in-progress" {
					extra = append(extra, l)
				}
			}
			suffix := ""
			if len(extra) > 0 {
				suffix = fmt.Sprintf(" (%s)", strings.Join(extra, ", "))
			}
			parts = append(parts, fmt.Sprintf("- #%d %s%s", int(num), title, suffix))
		}
	}

	if len(blocked) > 0 {
		parts = append(parts, "### Blocked")
		for _, i := range blocked {
			num, _ := i["number"].(float64)
			title, _ := i["title"].(string)
			parts = append(parts, fmt.Sprintf("- #%d %s", int(num), title))
		}
	}

	if len(ready) > 0 {
		parts = append(parts, "### Ready")
		limit := ready
		if len(limit) > 5 {
			limit = limit[:5]
		}
		for _, i := range limit {
			num, _ := i["number"].(float64)
			title, _ := i["title"].(string)
			labels := pmLabelNames(i)
			suffix := ""
			if len(labels) > 0 {
				suffix = fmt.Sprintf(" (%s)", strings.Join(labels, ", "))
			}
			parts = append(parts, fmt.Sprintf("- #%d %s%s", int(num), title, suffix))
		}
		if len(ready) > 5 {
			parts = append(parts, fmt.Sprintf("  ... and %d more", len(ready)-5))
		}
	}

	parts = append(parts, fmt.Sprintf(
		"**Total**: %d open (%d in-progress, %d blocked, %d ready)",
		len(issues), len(inProgress), len(blocked), len(ready),
	))

	// Branch context
	branchCtx := pmBranchContext(repo)
	if branchCtx != "" {
		parts = append(parts, "", branchCtx)
	}

	return core.Message(strings.Join(parts, "\n"))
}

func pmBranchContext(repo string) string {
	r := core.RunCmd([]string{"git", "branch", "--show-current"}, "", 3*time.Second, "")
	if r == nil || r.ExitCode != 0 {
		return ""
	}
	branch := strings.TrimSpace(r.Stdout)
	m := pmBranchIssueRe.FindStringSubmatch(branch)
	if m == nil {
		return ""
	}
	issueNum := m[1]

	out, err := clients.RunGH([]string{
		"issue", "view", issueNum, "--repo", repo, "--json", "title,body,state",
	})
	if err != nil {
		return fmt.Sprintf("### Current: #%s\nOn branch `%s`, commits auto-sync to #%s.", issueNum, branch, issueNum)
	}

	var data map[string]any
	if err := json.Unmarshal([]byte(out), &data); err != nil {
		return fmt.Sprintf("### Current: #%s\nOn branch `%s`, commits auto-sync to #%s.", issueNum, branch, issueNum)
	}

	title, _ := data["title"].(string)
	body, _ := data["body"].(string)

	var criteria []string
	for _, line := range strings.Split(body, "\n") {
		if regexp.MustCompile(`^\s*-\s*\[[ xX]\]`).MatchString(line) {
			criteria = append(criteria, strings.TrimSpace(line))
		}
	}

	lines := []string{fmt.Sprintf("### Current: #%s %s", issueNum, title)}
	lines = append(lines, fmt.Sprintf("On branch `%s`, commits auto-sync to #%s.", branch, issueNum))
	if len(criteria) > 0 {
		lines = append(lines, "Acceptance criteria:")
		if len(criteria) > 10 {
			criteria = criteria[:10]
		}
		lines = append(lines, criteria...)
	}
	return strings.Join(lines, "\n")
}

// ---------------------------------------------------------------------------
// B. PostToolUse/Bash
// ---------------------------------------------------------------------------

func pmPostBash(toolInput map[string]any, repo string) core.HookResult {
	command, _ := toolInput["command"].(string)
	if strings.Contains(command, "git commit") {
		return pmHandleCommit(repo)
	}
	if strings.Contains(command, "git merge") {
		return pmHandleMerge(command, repo)
	}
	if strings.Contains(command, "git worktree remove") {
		return pmHandleWorktreeRemove(command, repo)
	}
	return core.Allow()
}

func pmHandleCommit(repo string) core.HookResult {
	logR := core.RunCmd([]string{"git", "log", "-1", "--format=%h%n%s"}, "", 3*time.Second, "")
	if logR == nil || logR.ExitCode != 0 {
		return core.Allow()
	}
	logLines := strings.SplitN(strings.TrimSpace(logR.Stdout), "\n", 2)
	if len(logLines) < 2 {
		return core.Allow()
	}
	commitHash := logLines[0]
	commitSubject := logLines[1]

	fullMsgR := core.RunCmd([]string{"git", "log", "-1", "--format=%s%n%b"}, "", 3*time.Second, "")
	fullMsg := commitSubject
	if fullMsgR != nil {
		fullMsg = strings.TrimSpace(fullMsgR.Stdout)
	}

	branchR := core.RunCmd([]string{"git", "branch", "--show-current"}, "", 3*time.Second, "")
	branch := ""
	if branchR != nil && branchR.ExitCode == 0 {
		branch = strings.TrimSpace(branchR.Stdout)
	}

	issueNums := pmExtractIssues(fullMsg, branch)
	if len(issueNums) == 0 {
		return core.Allow()
	}

	// Cap to MAX_SYNC_ISSUES
	capped := pmSortedCapped(issueNums, pmMaxSyncIssues)

	// Diff stat for comment body
	diffStat := ""
	diffR := core.RunCmd([]string{"git", "diff", "--stat", "HEAD~1..HEAD"}, "", 3*time.Second, "")
	if diffR != nil && diffR.ExitCode == 0 {
		statLines := strings.Split(strings.TrimSpace(diffR.Stdout), "\n")
		if len(statLines) > 0 {
			fileLines := statLines
			summaryLine := ""
			if len(statLines) > 1 {
				fileLines = statLines[:len(statLines)-1]
				summaryLine = statLines[len(statLines)-1]
			}
			if len(fileLines) > 8 {
				extra := len(fileLines) - 8
				fileLines = fileLines[:8]
				diffStat = "\n```\n" + strings.Join(fileLines, "\n") +
					fmt.Sprintf("\n  ... and %d more files", extra)
			} else {
				diffStat = "\n```\n" + strings.Join(fileLines, "\n")
			}
			if summaryLine != "" {
				diffStat += "\n" + summaryLine
			}
			diffStat += "\n```"
		}
	}

	for _, num := range capped {
		body := fmt.Sprintf("Commit `%s`: %s", commitHash, commitSubject)
		if diffStat != "" {
			body += "\n" + diffStat
		}
		clients.RunGHBackground([]string{
			"issue", "comment", num, "--repo", repo, "--body", body,
		})
	}

	labels := make([]string, len(capped))
	for i, n := range capped {
		labels[i] = "#" + n
	}
	extra := ""
	if len(issueNums) > pmMaxSyncIssues {
		extra = fmt.Sprintf(" (+%d skipped)", len(issueNums)-pmMaxSyncIssues)
	}
	return core.Message(fmt.Sprintf("Syncing commit %s to %s%s", commitHash, strings.Join(labels, ", "), extra))
}

func pmHandleMerge(command, repo string) core.HookResult {
	parts := strings.Fields(command)
	branchName := ""
	for i, p := range parts {
		if p == "merge" {
			for j := i + 1; j < len(parts); j++ {
				if !strings.HasPrefix(parts[j], "-") {
					branchName = parts[j]
					break
				}
			}
			break
		}
	}
	if branchName == "" {
		return core.Allow()
	}
	m := pmBranchIssueRe.FindStringSubmatch(branchName)
	if m == nil {
		return core.Allow()
	}
	issueNum := m[1]

	out, err := clients.RunGH([]string{
		"issue", "view", issueNum, "--repo", repo, "--json", "state,title",
	})
	if err != nil {
		return core.Allow()
	}
	var data map[string]any
	if err := json.Unmarshal([]byte(out), &data); err != nil {
		return core.Allow()
	}
	if data["state"] == "OPEN" {
		title, _ := data["title"].(string)
		return core.Message(fmt.Sprintf(
			"Merge detected for #%s (%s). Issue is still OPEN — consider closing it with: `gh issue close %s --repo %s`",
			issueNum, title, issueNum, repo,
		))
	}
	return core.Allow()
}

func pmHandleWorktreeRemove(command, repo string) core.HookResult {
	m := pmBranchIssueRe.FindStringSubmatch(command)
	if m == nil {
		return core.Allow()
	}
	issueNum := m[1]

	out, err := clients.RunGH([]string{
		"issue", "view", issueNum, "--repo", repo, "--json", "state",
	})
	if err != nil {
		return core.Allow()
	}
	var data map[string]any
	if err := json.Unmarshal([]byte(out), &data); err != nil {
		return core.Allow()
	}
	if data["state"] == "OPEN" {
		return core.Message(fmt.Sprintf("Worktree removed but #%s is still OPEN. Forgot to close it?", issueNum))
	}
	return core.Allow()
}

// ---------------------------------------------------------------------------
// C. Stop
// ---------------------------------------------------------------------------

func pmOnStop() core.HookResult {
	state := pmLoadState()
	if state == nil {
		return core.Allow()
	}
	inProgress, _ := state["in_progress"].([]any)
	ready, _ := state["ready"].([]any)
	if len(inProgress) == 0 && len(ready) > 0 {
		issues, _ := state["issues"].([]any)
		issueMap := map[float64]map[string]any{}
		for _, raw := range issues {
			if issue, ok := raw.(map[string]any); ok {
				if num, ok := issue["number"].(float64); ok {
					issueMap[num] = issue
				}
			}
		}
		nextNum, _ := ready[0].(float64)
		if issue, ok := issueMap[nextNum]; ok {
			title, _ := issue["title"].(string)
			return core.Message(fmt.Sprintf("No in-progress issues. Next candidate: #%d %s", int(nextNum), title))
		}
	}
	return core.Allow()
}

// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------

func pmExtractIssues(text, branch string) []string {
	seen := map[string]bool{}
	// Closes/Fixes/Refs/Part of #N
	re1 := regexp.MustCompile(`(?:Closes|Fixes|Refs|Part of)\s+#(\d+)`)
	for _, m := range re1.FindAllStringSubmatch(text, -1) {
		seen[m[1]] = true
	}
	// Standalone #N
	re2 := regexp.MustCompile(`(?:^|[^\w])#(\d+)\b`)
	for _, m := range re2.FindAllStringSubmatch(text, -1) {
		seen[m[1]] = true
	}
	// Branch
	if branch != "" {
		if m := pmBranchIssueRe.FindStringSubmatch(branch); m != nil {
			seen[m[1]] = true
		}
	}
	result := make([]string, 0, len(seen))
	for k := range seen {
		result = append(result, k)
	}
	return result
}

func pmSortedCapped(nums []string, max int) []string {
	// Simple insertion sort (small N)
	sorted := make([]string, len(nums))
	copy(sorted, nums)
	for i := 1; i < len(sorted); i++ {
		for j := i; j > 0 && sorted[j] < sorted[j-1]; j-- {
			sorted[j], sorted[j-1] = sorted[j-1], sorted[j]
		}
	}
	if len(sorted) > max {
		return sorted[:max]
	}
	return sorted
}

func pmLabelNames(issue map[string]any) []string {
	labels, _ := issue["labels"].([]any)
	names := make([]string, 0, len(labels))
	for _, l := range labels {
		if lb, ok := l.(map[string]any); ok {
			if name, ok := lb["name"].(string); ok {
				names = append(names, name)
			}
		}
	}
	return names
}

func pmHasLabel(labels []string, target string) bool {
	for _, l := range labels {
		if l == target {
			return true
		}
	}
	return false
}

func pmNumbers(issues []map[string]any) []float64 {
	nums := make([]float64, 0, len(issues))
	for _, i := range issues {
		if n, ok := i["number"].(float64); ok {
			nums = append(nums, n)
		}
	}
	return nums
}

func pmLoadState() map[string]any {
	data, err := os.ReadFile(pmStateFile)
	if err != nil {
		return nil
	}
	var state map[string]any
	if err := json.Unmarshal(data, &state); err != nil {
		return nil
	}
	return state
}

func pmSaveState(data map[string]any) {
	b, err := json.Marshal(data)
	if err != nil {
		return
	}
	_ = os.WriteFile(pmStateFile, b, 0o644)
}

File: stations/_archive/hook-observatory-py/handlers/pm_autopilot.py
"""
PM Auto-Pilot — unified handler for SessionStart, PostToolUse/Bash, Stop.

Replaces manual /pm:* workflow with automatic:
  - SessionStart: inject open issues summary + current branch context
  - PostToolUse/Bash: enhanced commit sync, merge close suggestion, worktree orphan detection
  - Stop: suggest next task if nothing in-progress

Non-blocking design:
  - SessionStart: sync (must return message) but tight timeouts (max ~13s)
  - PostToolUse: git reads sync (fast), gh writes fire-and-forget background
  - Stop: local file read only (<1ms)
"""

from __future__ import annotations

import json
import re

from .base import ALLOW, HookResult, message, run_background, run_cmd
from .hook_config import cfg

REPO = cfg.get("github", {}).get("repo", "")
STATE_FILE = "/tmp/pm-autopilot-state.json"  # noqa: S108
MAX_SYNC_ISSUES = 3


# ---------------------------------------------------------------------------
# Main dispatcher
# ---------------------------------------------------------------------------


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    if not REPO:
        return ALLOW
    if event_type == "SessionStart":
        return _session_start()
    if event_type == "PostToolUse" and tool_name == "Bash":
        return _post_bash(tool_input)
    if event_type == "Stop":
        return _on_stop()
    return ALLOW


# ---------------------------------------------------------------------------
# A. SessionStart — environment awareness injection (sync, tight timeouts)
# ---------------------------------------------------------------------------


def _session_start() -> HookResult:
    # 1. Fetch open issues (must be sync for message injection)
    result = run_cmd(
        [
            "gh",
            "issue",
            "list",
            "--state",
            "open",
            "--json",
            "number,title,labels",
            "--limit",
            "20",
            "--repo",
            REPO,
        ],
        timeout=5,
    )
    if result is None or result.returncode != 0:
        _save_pm_state({"issues": [], "in_progress": [], "ready": []})
        return ALLOW

    try:
        issues = json.loads(result.stdout.strip())
    except (json.JSONDecodeError, ValueError):
        _save_pm_state({"issues": [], "in_progress": [], "ready": []})
        return ALLOW

    if not issues:
        _save_pm_state({"issues": [], "in_progress": [], "ready": []})
        return ALLOW

    # 2. Categorize by label
    in_progress = []
    blocked = []
    ready = []
    for issue in issues:
        label_names = [lb["name"] for lb in issue.get("labels", [])]
        if "in-progress" in label_names:
            in_progress.append(issue)
        elif "blocked" in label_names:
            blocked.append(issue)
        else:
            ready.append(issue)

    # Save state for Stop handler
    _save_pm_state(
        {
            "issues": issues,
            "in_progress": [i["number"] for i in in_progress],
            "ready": [i["number"] for i in ready],
            "blocked": [i["number"] for i in blocked],
        }
    )

    # 3. Build markdown summary
    parts = ["## GitHub PM Status"]

    if in_progress:
        parts.append("### In Progress")
        for i in in_progress:
            labels = ", ".join(
                lb["name"] for lb in i.get("labels", []) if lb["name"] != "in-progress"
            )
            suffix = f" ({labels})" if labels else ""
            parts.append(f"- #{i['number']} {i['title']}{suffix}")

    if blocked:
        parts.append("### Blocked")
        for i in blocked:
            parts.append(f"- #{i['number']} {i['title']}")

    if ready:
        parts.append("### Ready")
        for i in ready[:5]:
            labels = ", ".join(lb["name"] for lb in i.get("labels", []))
            suffix = f" ({labels})" if labels else ""
            parts.append(f"- #{i['number']} {i['title']}{suffix}")
        if len(ready) > 5:
            parts.append(f"  ... and {len(ready) - 5} more")

    parts.append(
        f"**Total**: {len(issues)} open"
        f" ({len(in_progress)} in-progress, {len(blocked)} blocked, {len(ready)} ready)"
    )

    # 4. Current branch context
    branch_ctx = _get_branch_context()
    if branch_ctx:
        parts.append("")
        parts.append(branch_ctx)

    return message("\n".join(parts))


def _get_branch_context() -> str:
    """If current branch contains #N, fetch issue details."""
    branch_result = run_cmd(["git", "branch", "--show-current"], timeout=3)
    if not branch_result or branch_result.returncode != 0:
        return ""

    branch = branch_result.stdout.strip()
    if not branch:
        return ""

    match = re.search(r"#(\d+)", branch)
    if not match:
        return ""

    issue_num = match.group(1)

    # Fetch issue body for acceptance criteria
    view_result = run_cmd(
        ["gh", "issue", "view", issue_num, "--repo", REPO, "--json", "title,body,state"],
        timeout=5,
    )
    if not view_result or view_result.returncode != 0:
        return (
            f"### Current: #{issue_num}\nOn branch `{branch}`, commits auto-sync to #{issue_num}."
        )

    try:
        data = json.loads(view_result.stdout.strip())
    except (json.JSONDecodeError, ValueError):
        return (
            f"### Current: #{issue_num}\nOn branch `{branch}`, commits auto-sync to #{issue_num}."
        )

    title = data.get("title", "")
    body = data.get("body", "") or ""

    # Extract acceptance criteria (lines starting with - [ ] or - [x])
    criteria = [line.strip() for line in body.splitlines() if re.match(r"^\s*-\s*\[[ xX]\]", line)]

    lines = [f"### Current: #{issue_num} {title}"]
    lines.append(f"On branch `{branch}`, commits auto-sync to #{issue_num}.")
    if criteria:
        lines.append("Acceptance criteria:")
        lines.extend(criteria[:10])

    return "\n".join(lines)


# ---------------------------------------------------------------------------
# B. PostToolUse/Bash — commit sync + merge detection + worktree orphan
#    gh writes are fire-and-forget (run_background), git reads are sync (fast)
# ---------------------------------------------------------------------------


def _post_bash(tool_input: dict) -> HookResult:
    command = tool_input.get("command", "")

    if "git commit" in command:
        return _handle_commit()
    if "git merge" in command:
        return _handle_merge(command)
    if "git worktree remove" in command:
        return _handle_worktree_remove(command)

    return ALLOW


def _handle_commit() -> HookResult:
    """Enhanced commit sync: post comment with diff stat to referenced issues.

    Git reads are sync (< 100ms each). gh issue comment is fire-and-forget.
    """
    # Get latest commit info
    log_result = run_cmd(["git", "log", "-1", "--format=%h%n%s"], timeout=3)
    if log_result is None or log_result.returncode != 0:
        return ALLOW

    log_lines = log_result.stdout.strip().split("\n", 1)
    if len(log_lines) < 2:
        return ALLOW

    commit_hash = log_lines[0]
    commit_subject = log_lines[1]

    # Get full commit message for issue extraction
    full_msg_result = run_cmd(["git", "log", "-1", "--format=%s%n%b"], timeout=3)
    full_msg = full_msg_result.stdout.strip() if full_msg_result else commit_subject

    # Extract issue numbers from commit message + branch
    branch_result = run_cmd(["git", "branch", "--show-current"], timeout=3)
    branch = branch_result.stdout.strip() if branch_result and branch_result.returncode == 0 else ""

    issue_numbers = _extract_issues(full_msg, branch)
    if not issue_numbers:
        return ALLOW

    # Cap to avoid runaway background processes
    capped = sorted(issue_numbers)[:MAX_SYNC_ISSUES]

    # Get diff stat for enhanced comment
    diff_result = run_cmd(["git", "diff", "--stat", "HEAD~1..HEAD"], timeout=3)
    diff_stat = ""
    if diff_result and diff_result.returncode == 0:
        stat_lines = diff_result.stdout.strip().splitlines()
        if stat_lines:
            file_lines = stat_lines[:-1][:8]
            summary_line = stat_lines[-1] if stat_lines else ""
            if file_lines:
                diff_stat = "\n```\n" + "\n".join(file_lines)
                if len(stat_lines) - 1 > 8:
                    diff_stat += f"\n  ... and {len(stat_lines) - 1 - 8} more files"
                if summary_line:
                    diff_stat += f"\n{summary_line}"
                diff_stat += "\n```"

    # Fire-and-forget: post comment on each referenced issue (background)
    for num in capped:
        comment_body = f"Commit `{commit_hash}`: {commit_subject}"
        if diff_stat:
            comment_body += f"\n{diff_stat}"
        run_background(
            ["gh", "issue", "comment", num, "--repo", REPO, "--body", comment_body],
        )

    labels = ", ".join(f"#{n}" for n in capped)
    extra = (
        f" (+{len(issue_numbers) - MAX_SYNC_ISSUES} skipped)"
        if len(issue_numbers) > MAX_SYNC_ISSUES
        else ""
    )
    return message(f"Syncing commit {commit_hash} to {labels}{extra}")


def _handle_merge(command: str) -> HookResult:
    """After git merge, suggest closing the issue if branch contains #N.

    gh issue view is fire-and-forget — writes suggestion to tmp for display.
    """
    parts = command.split()
    branch_name = ""
    for i, p in enumerate(parts):
        if p == "merge":
            for j in range(i + 1, len(parts)):
                if not parts[j].startswith("-"):
                    branch_name = parts[j]
                    break
            break

    if not branch_name:
        return ALLOW

    match = re.search(r"#(\d+)", branch_name)
    if not match:
        return ALLOW

    issue_num = match.group(1)

    # Check if issue is still open (sync but tight timeout)
    view_result = run_cmd(
        ["gh", "issue", "view", issue_num, "--repo", REPO, "--json", "state,title"],
        timeout=3,
    )
    if not view_result or view_result.returncode != 0:
        return ALLOW

    try:
        data = json.loads(view_result.stdout.strip())
    except (json.JSONDecodeError, ValueError):
        return ALLOW

    if data.get("state") == "OPEN":
        title = data.get("title", "")
        return message(
            f"Merge detected for #{issue_num} ({title}). "
            f"Issue is still OPEN — consider closing it with: "
            f"`gh issue close {issue_num} --repo {REPO}`"
        )

    return ALLOW


def _handle_worktree_remove(command: str) -> HookResult:
    """Warn if removing a worktree tied to an open issue."""
    match = re.search(r"#(\d+)", command)
    if not match:
        return ALLOW

    issue_num = match.group(1)

    view_result = run_cmd(
        ["gh", "issue", "view", issue_num, "--repo", REPO, "--json", "state"],
        timeout=3,
    )
    if not view_result or view_result.returncode != 0:
        return ALLOW

    try:
        data = json.loads(view_result.stdout.strip())
    except (json.JSONDecodeError, ValueError):
        return ALLOW

    if data.get("state") == "OPEN":
        return message(f"Worktree removed but #{issue_num} is still OPEN. Forgot to close it?")

    return ALLOW


# ---------------------------------------------------------------------------
# C. Stop — suggest next task (local file read only, <1ms)
# ---------------------------------------------------------------------------


def _on_stop() -> HookResult:
    state = _load_pm_state()
    if not state:
        return ALLOW

    in_progress = state.get("in_progress", [])
    ready = state.get("ready", [])

    if not in_progress and ready:
        issues = state.get("issues", [])
        issue_map = {i["number"]: i for i in issues}
        next_num = ready[0]
        next_issue = issue_map.get(next_num)
        if next_issue:
            title = next_issue.get("title", "")
            return message(f"No in-progress issues. Next candidate: #{next_num} {title}")

    return ALLOW


# ---------------------------------------------------------------------------
# Shared utilities
# ---------------------------------------------------------------------------


def _extract_issues(text: str, branch: str) -> set[str]:
    """Extract issue numbers from commit message text and branch name."""
    nums: set[str] = set()

    # Match: Closes #N, Fixes #N, Refs #N, Part of #N
    for m in re.finditer(r"(?:Closes|Fixes|Refs|Part of)\s+#(\d+)", text):
        nums.add(m.group(1))
    # Match standalone #N (not preceded by word char — avoids color hex)
    for m in re.finditer(r"(?<!\w)#(\d+)\b", text):
        nums.add(m.group(1))

    # Branch name: feature/slug-#42
    if branch:
        branch_match = re.search(r"#(\d+)", branch)
        if branch_match:
            nums.add(branch_match.group(1))

    return nums


def _load_pm_state() -> dict:
    try:
        with open(STATE_FILE) as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError, ValueError):
        return {}


def _save_pm_state(data: dict) -> None:
    try:
        with open(STATE_FILE, "w") as f:
            json.dump(data, f)
    except OSError:
        pass
